### PR TITLE
PUBDEV-8971: Speed-up XGBoost scoring on wide datasets

### PR DESCRIPTION
**Pull request by michalkurka**
The goal of this PR is to speed-up XGBoost scoring, especially on wide datasets and in cases when only a portion of training columns is used to by the model.

Idea: It is expensive to repeatedly (=for each scoring) convert H2O Frame into float[] rows for Predictor scoring. In this change we only convert the columns that are actually actively used by the model.

Measurements (only measures scoring time during training, time to build trees is ignored)

Case MNIST, 60k rows, 785 cols

    Default settings + score_each_iteration = True

    Old: 15.7s
    New: 9.4s

Syntetic, 300k rows, 800 cols

    Extreme settings - max_depth = 1, ntrees = 10 (this forces a small model - only up to 10 different features can be used by the model)

    Old: 89s
    New: 8.1s (10x+ speed-up)

This is an early preview of the implementation demonstrating the direction I would like to evolve the PR in.

### CODE REVIEW

**Comment by mn-mikke** [comment]
Thank @michalkurka, your initiative is highly appreciated!

**Comment by michalkurka** [comment]
thank @mn-mikke, I think this is now ready for review

**Comment by valenad1** [comment]
Test passed 👍

**Comment by michalkurka** [comment]
@valenad1 @mn-mikke thanks for running the tests and confirming they pass; FIY - this PR is a prerequisite to https://github.com/h2oai/h2o-3/pull/6481

File: h2o-extensions/xgboost/src/main/java/biz/k11i/xgboost/tree/NodeHelper.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * These methods can eventually be folded back into the original class.
+ */
+public class NodeHelper {
+
+    public static RegTreeNode read(ModelReader reader) throws IOException {
+        return new RegTreeImpl.Node(reader);
+    }
+
+    public static boolean isEqual(RegTreeNode left, RegTreeNode right) {
+        return left.getParentIndex() == right.getParentIndex()
```
Comment (mn-mikke) [inline h2o-extensions/xgboost/src/main/java/biz/k11i/xgboost/tree/NodeHelper.java:18]: Shouldn't this method also expect that `left` or `right` could be `null`?

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/predict/XGBoostJavaBigScorePredictTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+@RunWith(H2ORunner.class)
+public class XGBoostJavaBigScorePredictTest extends TestUtil  {
+
+    @Test
+    public void testFindUsedColumns_few() {
+        XGBoostModel.XGBoostParameters parms = new XGBoostModel.XGBoostParameters();
+        parms._weights_column = "CAPSULE";
+
+        boolean[] used = checkFindUsedColumns(parms, 5);
+        assertTrue(countTrues(used) > 0);
```
Comment (mn-mikke) [inline h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/predict/XGBoostJavaBigScorePredictTest.java:25]: Is there any logical explanation why these two conditions must hold?

File: h2o-extensions/xgboost/src/main/java/biz/k11i/xgboost/tree/NodeHelper.java (hunk 1 above)
Comment (michalkurka) [reply h2o-extensions/xgboost/src/main/java/biz/k11i/xgboost/tree/NodeHelper.java:18]: yes, in this context null, can be considered - I will add that

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/predict/XGBoostJavaBigScorePredictTest.java (hunk 2 above)
Comment (michalkurka) [reply h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/predict/XGBoostJavaBigScorePredictTest.java]: I improved the test to be more clear, please let me know if it makes sense this way now.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/predict/XGBoostJavaBigScoreChunkPredict.java [hunk 3]
```diff
@@ -16,11 +16,13 @@ public class XGBoostJavaBigScoreChunkPredict implements XGBoostPredict, Model.Bi
...
   private final MutableOneHotEncoderFVec _row;
   private final int _offsetIndex;
+  private final boolean[] _usedColumns;
 
   public XGBoostJavaBigScoreChunkPredict(
       DataInfo di, XGBoostOutput output,
       XGBoostModel.XGBoostParameters parms,
-      double threshold, Predictor predictor,
+      double threshold,
+      Predictor predictor, boolean[] usedColumns,
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/predict/XGBoostJavaBigScoreChunkPredict.java]: fixed

File: h2o-extensions/xgboost/src/main/java/biz/k11i/xgboost/tree/NodeHelper.java (hunk 1 above)
Comment (michalkurka) [reply h2o-extensions/xgboost/src/main/java/biz/k11i/xgboost/tree/NodeHelper.java]: fixed

**Review by mn-mikke — APPROVED** [review]
LGTM, thank you a lot!

**Comment by michalkurka** [comment]
Can you please re-run the PR metadata check or merge the PR for me?

**Comment by mn-mikke** [comment]
I've re-run the failed tests:

There was one related to external xgboost:
```
versionFromGradle='3.39.0',projectVersion='3.39.0.99999',branch='(HEAD detached at c1b878a)',lastCommitHash='c1b878a95bdaef4814f0ea8bde5c140fbda61719',gitDescribe='jenkins-master-6095-3-gc1b878a',compiledOn='2023-02-02 19:47:57',compiledBy='jenkins'
loading the following test utility modules:  ['/home/jenkins/slave_dir_from_mr-0xc1/workspace/h2o-3-pipeline_PR-6469@2/external-xgboost/h2o-3/h2o-py/tests/_utils.py', '/home/jenkins/slave_dir_from_mr-0xc1/workspace/h2o-3-pipeline_PR-6469@2/external-xgboost/h2o-3/h2o-py/tests/testdir_algos/automl/_automl_utils.py', '/home/jenkins/slave_dir_from_mr-0xc1/workspace/h2o-3-pipeline_PR-6469@2/external-xgboost/h2o-3/h2o-py/tests/testdir_algos/automl/leaderboard/_leaderboard_utils.py']
[2023-02-02 20:14:56] Connect to h2o on IP: 127.0.0.1 PORT: 40006

Not using any auth
Parse progress: |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| (done) 100%
xgboost Model Build progress: |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ (failed)
Traceback (most recent call last):
  File "/home/jenkins/slave_dir_from_mr-0xc1/workspace/h2o-3-pipeline_PR-6469@2/external-xgboost/h2o-3/h2o-py/scripts/h2o-py-test-setup.py", line 186, in <module>
    h2o_test_setup(sys.argv)
  File "/home/jenkins/slave_dir_from_mr-0xc1/workspace/h2o-3-pipeline_PR-6469@2/external-xgboost/h2o-3/h2o-py/scripts/h2o-py-test-setup.py", line 181, in h2o_test_setup
    elif _IS_PYUNIT_:    pyunit_utils.pyunit_exec(_TEST_NAME_)
  File "/home/jenkins/slave_dir_from_mr-0xc1/workspace/h2o-3-pipeline_PR-6469@2/external-xgboost/h2o-3/h2o-py/tests/pyunit_utils/utilsPY.py", line 686, in pyunit_exec
    exec(pyunit_c, dict(__name__='__main__', __file__=test_path))  # forcing module name to ensure that the test behaves the same way as when executed using `python my_test.py`
  File "/home/jenkins/slave_dir_from_mr-0xc1/workspace/h2o-3-pipeline_PR-6469@2/external-xgboost/h2o-3/h2o-py/tests/testdir_algos/xgboost/pyunit_xgboost_eval_metric_cv_large.py", line 59, in <module>
    pyunit_utils.standalone_test(test_gam_model_predict)
  File "/home/jenkins/slave_dir_from_mr-0xc1/workspace/h2o-3-pipeline_PR-6469@2/external-xgboost/h2o-3/h2o-py/tests/pyunit_utils/utilsPY.py", line 701, in standalone_test
    test()
  File "/home/jenkins/slave_dir_from_mr-0xc1/workspace/h2o-3-pipeline_PR-6469@2/external-xgboost/h2o-3/h2o-py/tests/testdir_algos/xgboost/pyunit_xgboost_eval_metric_cv_large.py", line 25, in test_gam_model_predict
    xgb_classic.train(y=y, training_frame=loan_df)
  File "/envs/h2o_env_python3.7/lib/python3.7/site-packages/h2o/estimators/estimator_base.py", line 108, in train
    self._train(parms, verbose=verbose)
  File "/envs/h2o_env_python3.7/lib/python3.7/site-packages/h2o/estimators/estimator_base.py", line 200, in _train
    job.poll(poll_updates=self._print_model_scoring_history if verbose else None)
  File "/envs/h2o_env_python3.7/lib/python3.7/site-packages/h2o/job.py", line 91, in poll
    "\n{}".format(self.job_key, self.exception, self.job["stacktrace"]))
OSError: Job with key $03017f000001479cffffffff$_81e08775ec9c46e0ac26bb54b7ff4698 failed with an exception: java.lang.RuntimeException: Error while training XGBoost model
stacktrace: 
java.lang.RuntimeException: Error while training XGBoost model
	at hex.tree.xgboost.XGBoost$XGBoostDriver.buildModel(XGBoost.java:429)
	at hex.tree.xgboost.XGBoost$XGBoostDriver.computeImpl(XGBoost.java:379)
	at hex.ModelBuilder$Driver.compute2(ModelBuilder.java:253)
	at water.H2O$H2OCountedCompleter.compute(H2O.java:1677)
	at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)
	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)
	at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:976)
	at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1479)
	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)
Caused by: DistributedException from /127.0.0.1:40010: 'Unexpected response (status: HTTP/1.1 500 Server Error).', caused by java.lang.IllegalStateException: Unexpected response (status: HTTP/1.1 500 Server Error).
	at water.MRTask.getResult(MRTask.java:660)
	at water.MRTask.getResult(MRTask.java:670)
	at water.MRTask.doAll(MRTask.java:555)
	at water.MRTask.doAllNodes(MRTask.java:568)
	at hex.tree.xgboost.task.AbstractXGBoostTask.run(AbstractXGBoostTask.java:45)
	at hex.tree.xgboost.exec.RemoteXGBoostExecutor.uploadMatrix(RemoteXGBoostExecutor.java:81)
	at hex.tree.xgboost.exec.RemoteXGBoostExecutor.<init>(RemoteXGBoostExecutor.java:53)
	at hex.tree.xgboost.XGBoost$XGBoostDriver.makeExecutor(XGBoost.java:393)
	at hex.tree.xgboost.XGBoost$XGBoostDriver.buildModel(XGBoost.java:425)
	... 8 more
Caused by: java.lang.IllegalStateException: Unexpected response (status: HTTP/1.1 500 Server Error).
	at hex.tree.xgboost.exec.XGBoostHttpClient.executeRequestAndReturnResponse(XGBoostHttpClient.java:224)
	at hex.tree.xgboost.exec.XGBoostHttpClient.uploadMatrixData(XGBoostHttpClient.java:203)
	at hex.tree.xgboost.task.XGBoostUploadMatrixTask$UploadSparseMatrixFun.map(XGBoostUploadMatrixTask.java:373)
	at water.LocalMR.compute2(LocalMR.java:89)
	... 6 more

Closing connection _sid_a24c at exit
```

**Comment by mn-mikke** [comment]
and also one multinode xgboost: 
```
1) testScoringTreesHaveCorrectNodesInfo(hex.tree.xgboost.predict.PredictorTest)
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at hex.tree.xgboost.predict.PredictorTest.assertNodesEqual(PredictorTest.java:75)
	at hex.tree.xgboost.predict.PredictorTest.testScoringTreesHaveCorrectNodesInfo(PredictorTest.java:64)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at water.runner.H2ORunnerBefores.evaluate(H2ORunnerBefores.java:30)
	at water.TestUtil$3$TimerStatement.evaluate(TestUtil.java:453)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at water.runner.H2ORunner.leaf(H2ORunner.java:92)
	at water.runner.H2ORunner.runChild(H2ORunner.java:61)
	at water.runner.H2ORunner.runChild(H2ORunner.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at water.runner.H2ORunnerAfters.evaluate(H2ORunnerAfters.java:32)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:105)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:94)
	at water.junit.H2OTestRunner.run(H2OTestRunner.java:43)
	at water.junit.H2OTestRunner.main(H2OTestRunner.java:54)

```

**Comment by mn-mikke** [comment]
@michalkurka The assert https://github.com/h2oai/h2o-3/pull/6469#issuecomment-1414492159 is still failing.

**Comment by michalkurka** [comment]
@mn-mikke thank you, I will look into it

**Comment by mn-mikke** [comment]
@michalkurka No, the external xgboost test failed only once. Just the  multi-node test failed again.

**Comment by mn-mikke** [comment]
Thanks a lot for your contribution. Merging the PR.

**Pull request merged**